File: AndroidApp/app/src/main/java/com/bitzquad/spindlydemo/MainActivity.java
package com.bitzquad.spindlydemo;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;

import com.google.androidbrowserhelper.trusted.LauncherActivity;
import com.google.gson.Gson;

public class MainActivity extends LauncherActivity {

    private long onCreateTime;


    // ------------- Activity ------------- //

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        onCreateTime = System.currentTimeMillis();

        if (checkAllPermissions()) {
            refresh();

        } else {
            openPermActivity();
        }

    }

    private void refresh() {

    }


    @Override
    public void onDestroy() {
        super.onDestroy();

        long currentTime = System.currentTimeMillis();
        if (onCreateTime + 3000 > currentTime) {
            Log.d("MainAct", "Suppressing onDestroy()");
            return;
        }

    }


    @Override
    public void onStop() {
        super.onStop();
    }

    @Override
    protected void onResume() {
        super.onResume();
        initApp();
    }

    private boolean firstRun = true;

    private void initApp() {
        if (firstRun) {
            firstRun = false;

            setAppName("Spindly Demo Java");
            setAppName("Presibit");
        }

        startSpindly();
    }

    void openPermActivity() {
        Intent intent = new Intent(this, StartupActivity.class);
        startActivity(intent);
    }

    boolean checkAllPermissions() {
        return true;
    }


    // ------------- Scan : Spindly Exports handling ------------- //

    private void startSpindly() {
        SpindlyUtils.initializeStatic(getCacheDir().getAbsolutePath(), getAssets());
        this.startService(new Intent(this, SpindlyService.class));
    }

    private void setAppName(String value) {
        Thread UIThread = new Thread() {
            @Override
            public void run() {
                spindlyapp.NamedExportedHubs UI = spindlyapp.Spindlyapp.namedExports();
                UI.getGlobal().getAppName().setValue(ToJSON(value));
            }
        };

        UIThread.start();
    }

    private String ToJSON(Object obj) {
        Gson gson = new Gson();
        return gson.toJson(obj);
    }

    public <T> T FromJSON(String json, Class<T> type) {
        Gson gson = new Gson();
        return gson.fromJson(json, type);
    }

}

File: AndroidApp/app/src/main/java/com/bitzquad/spindlydemo/MainActivity.kt
package com.bitzquad.spindlydemo

import android.content.Intent
import android.util.Log
import java.io.*


class MainActivity : com.google.androidbrowserhelper.trusted.LauncherActivity() {

    override fun onResume() {
        super.onResume()

        val cwd = cacheDir.absolutePath + "/"
        copyFileOrDir("WebApp", cwd)
        startSpindlyOnAnotherThread()

    }

    private fun startSpindlyOnAnotherThread() {

        Intent(this, SpindlyService::class.java).also { intent ->
            startService(intent)
        }

//
//        val goThread = Thread {
//            GoApp.mainMobile(cwd)
//        }
//        goThread.start()
    }

    private fun copyFileOrDir(path: String, dest: String) {

        val assetManager = this.assets
        val assets: Array<String>?
        try {
            assets = assetManager.list(path)
            if (assets!!.isEmpty()) {
                copyFile(path, dest)
            } else {
                val fullPath = dest + path
                val dir = File(fullPath)
                if (!dir.exists()) dir.mkdir()
                for (i in assets.indices) {
                    copyFileOrDir(path + "/" + assets[i], dest)
                }
            }
        } catch (ex: IOException) {
            Log.e("tag", "I/O Exception", ex)
        }
    }

    private fun copyFile(filename: String, dest: String) {


        val assetManager = this.assets
        val inputStrm: InputStream?
        val out: OutputStream?
        try {
            inputStrm = assetManager.open(filename)
            val newFileName = dest + filename

            Log.d("Copy", "Copy $filename to $newFileName")

            out = FileOutputStream(newFileName)
            val buffer = ByteArray(1024)
            var read: Int
            while (inputStrm.read(buffer).also { read = it } != -1) {
                out.write(buffer, 0, read)
            }
            inputStrm.close()
            out.flush()
            out.close()
        } catch (e: Exception) {
            Log.e("tag", e.message!!)
        }
    }
}

